streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java

```diff
@@ -490,6 +490,31 @@ public void shouldTimeIteratorDuration() {
         assertThat((double) iteratorDurationMaxMetric.metricValue(), equalTo(3.0 * TimeUnit.MILLISECONDS.toNanos(1)));
     }
 
+    @Test
+    public void shouldTrackOldestOpenIteratorTimestamp() {
+        when(inner.all()).thenReturn(KeyValueIterators.emptyIterator());
+        init();
+
+        final KafkaMetric oldestIteratorTimestampMetric = metric("oldest-iterator-open-since-ms");
+        assertThat(oldestIteratorTimestampMetric, not(nullValue()));
+
+        assertThat(oldestIteratorTimestampMetric.metricValue(), nullValue());
+
+        try (final KeyValueIterator<String, String> first = metered.all()) {
+            final long oldestTimestamp = mockTime.milliseconds();
+            assertThat((Long) oldestIteratorTimestampMetric.metricValue(), equalTo(oldestTimestamp));
+            mockTime.sleep(100);
+
+            // open a second iterator before closing the first to test that we still produce the first iterator's timestamp
+            try (final KeyValueIterator<String, String> second = metered.all()) {
+                assertThat((Long) oldestIteratorTimestampMetric.metricValue(), equalTo(oldestTimestamp));
+                mockTime.sleep(100);
+            }
```

Comment:
It would be better to not close the second iterator here, but close the first one first, to see if the metric advances to the second's iterator create ts -- would need some rewriting of the test; try-with-resource won't allow for proper nesting, but we can still use try-finally.

Might actually be best, to open like 5 iterators and close them in some non-linear order (including closing the oldest one like 2 or 3 times) to verify correct behavior.

Reply:
I've updated the tests with your first suggestion. I didn't go further as I feel that this case is already adequately tested.